### PR TITLE
Fix verification step

### PR DESCRIPTION
**Pull request by tramuntanal**
This PR fixes the verification process where neither the `VerifyWoRegistrationForm` nor the `DoVerifyWoRegistration` command were actually telling `authorization_handlers` to verify authorization data.

With this fix, the command keeps the previous behavior where first checks if there's already an authorization. If an authorization exists, redirects back and tells the user to use the already verified account, if not, then tries to verify the user with the current data in the form.
A special case are impersonated (managed) users that are re-authenticated every time.

Note that the verification process is implemented in the `VerifyWoRegistrationForm#verify_against_enabled_authorization_handlers`, which delegates the task to the first existing handler of type direct.

Related to #2

### CODE REVIEW

**Comment by tramuntanal** [comment]
@mrcasals , @operez-ddgi would you like to review?

File: app/controllers/decidim/verify_wo_registration/verifications_controller.rb [hunk 1]
```diff
@@ -58,7 +60,7 @@ def component
 
       # Should we validate that only 'direct' verification handlers are enabled?
       def validate_verification_workflow_manifests!
-        return if all_verifications_of_type_direct?(component)
+        return if Decidim::VerifyWoRegistration::ApplicationHelper.verify_wo_registration_custom_modal?(component)
```
Comment (Leusev) [inline app/controllers/decidim/verify_wo_registration/verifications_controller.rb:63]: You've change this check by calling helper method
https://github.com/CodiTramuntana/decidim-verify_wo_registration/blob/e909116b44e5fa781d5ad260be70f472b27c815b/app/helpers/decidim/verify_wo_registration/application_helper.rb#L8
that makes double check, same as here before plus `component.settings.try(:supports_without_registration)` that's strong. Seems good for me :smile:

**Review by Leusev — APPROVED** [review]
It looks good for me @tramuntanal

File: app/controllers/decidim/verify_wo_registration/verifications_controller.rb (hunk 1 above)
Comment (tramuntanal) [reply app/controllers/decidim/verify_wo_registration/verifications_controller.rb:63]: yes, I added this check to disable the controller if the feature is not enabled but in fact the verification process is implemented in the `VerifyWoRegistrationForm#verify_against_enabled_authorization_handlers`.
I've updated the description of the PR

**Comment by diputaciodegirona** [comment]
I am afraid this fix is incomplete.
It should check if there's an existing authorization at "authorizations_exists?" but only for the current user data. I don't see any check on the submitted user data against the census...

**Comment by tramuntanal** [comment]
Hi @diputaciodegirona, `authorizations_exists?` only checks if there's a previous authorization for the current user and verification system.
The verification itself is done in `VerifyWoRegistrationForm#verify_against_enabled_authorization_handlers`, where the form delegates the verification process to the first handler of type "direct" (normally there's only one). Then it depends on the currently installed verification system how the verification is done.

In summary, this module delegates the verification step to the currently installed/configured verification system. The verification itself is done by exactly the same code where it is done when this module is not installed.

**Comment by diputaciodegirona** [comment]
Hi @tramuntanal ,

thank-you very much for the explanation!

File: app/commands/decidim/verify_wo_registration/do_verify_wo_registration.rb [hunk 2]
```diff
@@ -20,10 +20,20 @@ def initialize(form)
...
           if existing_registered_user?
+            # ignore form data, make the participant use the already existing user
             broadcast(:use_registered_user)
           elsif existing_impersonated_user?
-            authorize_user
-            broadcast(:ok, user)
+            # we can not reuse existing authorization because it will raise "A participant is already authorized with the same data." on @form.valid?
+            # as it is an impersonated user, we can safely destroy it and perform the verification process again
+            transaction do
+              destroy_authorization
```
Comment (mrcasals) [inline app/commands/decidim/verify_wo_registration/do_verify_wo_registration.rb:29]: If you destroy the authorization, then I can vote for the same proposal multiple times:

1. I use this module to verify myself without registering and get a 30min session. This creates a new user, say user ID 1
2. I vote for the proposal A. Vote is related to user ID 1.
3. I wait until the 30min session has expired.
4. I use this module to verify myself again. Since the verification is destroyed, I can verify myself again and a new user is created, say user ID 2.
5. Since the user is brand new, I can vote for the proposal A again. Vote is now related to user ID 2.

I'd remove this

**Review by mrcasals — CHANGES_REQUESTED** [review]

File: app/commands/decidim/verify_wo_registration/do_verify_wo_registration.rb (hunk 2 above)
Comment (tramuntanal) [reply app/commands/decidim/verify_wo_registration/do_verify_wo_registration.rb:29]: Good catch @mrcasals you're right!
Let me see if I can find a way to bind the new authorization to the previously impersonated user, or otherwise do not allow new sessions for previously impersonated users

Comment (tramuntanal) [reply app/commands/decidim/verify_wo_registration/do_verify_wo_registration.rb:29]: After reviewing the code, I'm not sure your proposed workflow is correct.
When we destroy the authorization on `destroy_authorization` the previously impersonated user (ID 1) is kept in the command as an attribute. Then, in `authorize_user` we do `create_or_update_authorizations`, we do not create a new impersonated user, we reassign the previously impersonated user (with ID 1) and no new user is created.

So the votes keeps being associated to new impersonated sessions.

Comment (tramuntanal) [reply app/commands/decidim/verify_wo_registration/do_verify_wo_registration.rb:29]: Although your workflow is may happen if the second workflow is as follows:
1. I use this module to verify myself without registering and get a 30min session. This creates a new user, say user ID 1
2. I vote for the proposal A. Vote is related to user ID 1.
3. An admin resets the authorizations in the middle of the voting period
4. All impersonated users (both via admin or using this module) will be able to verify and vote again via all means: registering, being impersonated in the admin or using this module.

Being said that, the behavior that this module implements is according the requirements of the module and with the principal philosophy of the module which is simplify the steps to the minimal for participants to be able to participate in processes that require being verified to increase participation.

So, as this is a corner case that can happen either the organization has this module enabled or not (so this may happen in any Decidim installation) I think the best will be to add a notice in the reset button to avoid admins reset authorizations in the middle of a process.

Comment (mrcasals) [reply app/commands/decidim/verify_wo_registration/do_verify_wo_registration.rb:29]: Ah, you're right, I missed the code in the `existing_impersonated_user?` method because since it ends in `?` I expected to only perform a boolean check, not to add extra logic to the process... 😕 

> An admin resets the authorizations in the middle of the voting period

You're right this can cause issues and this might be out of the scope of this module. Maybe update the README file with a "Warning" section explaining the possible consequences of resetting verifications in the middle of a voting process?

**Comment by mrcasals** [comment]
@tramuntanal I've checked the code and it looks good to me! The problem you express [here](https://github.com/CodiTramuntana/decidim-verify_wo_registration/pull/4#discussion_r566148523) already happens in Decidim and it's not this module's fault 😄

**Review by mrcasals — APPROVED** [review]

**Comment by tramuntanal** [comment]
Thanks @mrcasals for this last review. We're merging
